crates/mun_hir/src/package_defs/collector.rs

```diff
@@ -0,0 +1,200 @@
+use super::PackageDefs;
+use crate::arena::map::ArenaMap;
+use crate::ids::{
+    FunctionLoc, Intern, ItemDefinitionId, ModuleId, PackageId, StructLoc, TypeAliasLoc,
+};
+use crate::item_scope::ItemScope;
+use crate::item_tree::{
+    Function, ItemTree, ItemTreeId, LocalItemTreeId, ModItem, Struct, StructDefKind, TypeAlias,
+};
+use crate::module_tree::{LocalModuleId, ModuleTree};
+use crate::visibility::RawVisibility;
+use crate::{DefDatabase, FileId, Name, PerNs, Visibility};
+use std::sync::Arc;
+
+pub(super) fn collect(db: &dyn DefDatabase, package_id: PackageId) -> PackageDefs {
+    let mut collector = DefCollector {
+        db,
+        package_id,
+        modules: Default::default(),
+        module_tree: db.module_tree(package_id),
+    };
+    collector.collect();
+    collector.finish()
+}
+
+struct DefCollector<'db> {
+    db: &'db dyn DefDatabase,
+    package_id: PackageId,
+    modules: ArenaMap<LocalModuleId, ItemScope>,
+    module_tree: Arc<ModuleTree>,
+}
+
+impl<'t> DefCollector<'t> {
+    /// Collects all information and stores it in the instance
+    fn collect(&mut self) {
+        // Collect all definitions in each module
+        let module_tree = self.module_tree.clone();
+
+        fn collect_modules_recursive<'t>(
+            collector: &mut DefCollector<'t>,
+            module_id: LocalModuleId,
+            parent: Option<(Name, LocalModuleId)>,
+        ) {
+            // Insert an empty item scope for this module, this will be filled in.
+            collector.modules.insert(module_id, ItemScope::default());
+
+            // If there is a file associated with the module, collect all definitions from it
+            let module_data = &collector.module_tree[module_id];
+            if let Some(file_id) = module_data.file {
+                let item_tree = collector.db.item_tree(file_id);
+                ModCollector {
```

Comment:
This notation throws me off a little as it seems like you are doing something to a temporary that is then destroyed (and resulting data lost). In reality, probably the `ModCollector` is a wrapper around the `collector` that ensures the data is stored in `collector`.
Could we add a function to the type of `collector` that wraps does this instead?

Reply:
I introduced a variable. I hope that makes it a bit clearer.

Reply:
I don't know that it helps solves the issue I pointed out. To me, it's still unclear that you are actually operating on the `DefCollector` through the helper struct. If the goal was to use helper functions that operate on `DefCollector`s public API, I'd probably just do so using unassociated `pub fn`s instead of creating a helper struct that obfuscates the control flow.

I might be overlooking something; or it's a personal preference, so I'm not making an issue of this.

Reply:
I hear you. I tried a couple of approaches:

- I moved everything to seperate function but then the argument list becomes quite large and it feels like code duplication
- I moved the code you highlighted to a seperate function inside the DefCollector (still keeping the ModCollector) but it requires a lot of parameters that feel very out of context in the new function. I could move the entire if statement to a function but that feels just as out of place to me.
- I tried keeping it the same but only remove the collector from the struct, but then a lot of lifetimes no longer work which doesnt improve the readibility of the code.

Would it help if I renamed the `ModCollector` to `ModCollectionContext`?

Reply:
I think a context would work. You'd then pass that context to (unassocitaed) methods that use it?

Reply:
I renamed the `ModCollector` to `ModCollectorContext`, I then made all its functions unassociated but that seemed even more weird to me because there is a lot of duplication. So I kept the `impl` of the `ModCollectorContext`.

Reply:
Thanks for all of your effort trying to improve this! 🙂